store/src/java/com/zimbra/cs/mailbox/ContactAutoComplete.java

```diff
@@ -756,6 +756,35 @@ public void addMatchedContacts(String query, Map<String, ? extends Object> attrs
         }
     }
 
+    private List<String> extractAllowedEmailsListForAccount(Map<String, ?> attrs, String fullName) throws ServiceException {
+        List<String> allowedEmailsListInGalAutocomplete = new ArrayList<>();
+
+        Account account = Provisioning.getInstance().get(Key.AccountBy.name, fullName);
+        boolean isCosHidden = ("TRUE".equalsIgnoreCase(account.getCOS().getAttr(Provisioning.A_zimbraHideAliasesInGal)));
+        boolean isDomainHidden = ("TRUE".equalsIgnoreCase(account.getProvisioning().getDomain(account).getAttr(Provisioning.A_zimbraHideAliasesInGal)));
+
+        ZimbraLog.mailbox.info("Values for isCosHidden is %s and isDomain Hidden is %s", isCosHidden, isDomainHidden);
+
+        for(String emailKey: mEmailKeys) {
+            if (!emailKey.equals(Constants.PRIMARY_EMAIL)) {
+                //check if all aliases of an account need to be bypassed
+                if (account.isHideAliasesInGal() || isCosHidden || isDomainHidden) {
+                    break;
+                }
```

Comment:
Well, this seems like a bug to me.

You seem to be assuming that the first account email is the primary email.
It seems to be that the if:

- The first account email is an alias
- The new settings about hiding alias is turned on

then this first break statement ends up in the main account email (primary key) not being added to the allowed emails for the account.

Reply:
please check the below command and its output, this is for autocomplete from CLI.

```
zimbra@telus-qa-1:~$  zmprov ACG telus-qa-1.eng.zimbra.com test
# name uid=test,ou=people,dc=telus-qa-1,dc=eng,dc=zimbra,dc=com
createTimeStamp: 20240708054059Z
email: test@telus-qa-1.eng.zimbra.com
email2: alias1hidden@telus-qa-1.eng.zimbra.com
email3: alias2hidden@telus-qa-1.eng.zimbra.com
email4: alias1@telus-qa-1.eng.zimbra.com
email5: alias2@telus-qa-1.eng.zimbra.com
email6: alias1hidden2@telus-qa-1.eng.zimbra.com
email7: alias1hidden22@telus-qa-1.eng.zimbra.com
fullName: test
lastName: test
modifyTimeStamp: 20240709133650Z
objectClass: inetOrgPerson
objectClass: zimbraAccount
objectClass: amavisAccount
zimbraId: 35327976-42a8-49e1-949b-c0c5a316bebe

# name uid=testuser,ou=people,dc=telus-qa-1,dc=eng,dc=zimbra,dc=com
createTimeStamp: 20240710045942Z
email: testuser@telus-qa-1.eng.zimbra.com
fullName: testuser
lastName: testuser
modifyTimeStamp: 20240710045942Z
objectClass: inetOrgPerson
objectClass: zimbraAccount
objectClass: amavisAccount
zimbraId: e4bce526-160a-4839-84d1-a649e4b30a3e
```
Here email corresponds to PrimaryEmail and email2, email3... corresponds to aliases created in same order. I checked in the implementation for this, the results with tags are returned directly from Ldap here,
https://github.com/Zimbra/zm-mailbox/blob/a2bf0d89d48dc16def8989daa601e157bab62e8c/store/src/java/com/zimbra/cs/ldap/unboundid/UBIDLdapContext.java#L584

Also from the UI,  I have this state captured for Intellij debug mode and Ldap state,
![image](https://github.com/Zimbra/zm-mailbox/assets/104493060/6bae4ead-5582-4466-90d2-8ed19b633ecd)

These both seem to confirm that 'email' always corresponds to primary email and it will be always present.

Please validate this understanding, please suggest other scenarios to check if any.

Reply:
Well, the way I would test this is changing:

```java
        for(String emailKey: mEmailKeys) {
            if (!emailKey.equals(Constants.PRIMARY_EMAIL)) {
```

to:

```java
        for(String emailKey: mEmailKeys) {
            break;
            if (!emailKey.equals(Constants.PRIMARY_EMAIL)) {
```
and check what happens then.

That illustrates my point that in the first loop `break` might happen. That's unless somehow, the first email of the loop is enforced by some ldap magic to be the primary email.

My assumption is that:

- You create an account with name `test1@`
- You then add `alias1@` and `alias2@`.
- You somehow set `alias2@` to be the primary email.
- Ldap returns `test1@` (which it's now an alias) as the first email of the loop

.

Reply:
I checked how changePrimaryEmail works..
When trying to change existing alias as primary email
```
zimbra@telus-qa-1:~$ zmprov cpe test@telus-qa-1.eng.zimbra.com alias1@telus-qa-1.eng.zimbra.com
ERROR: account.ACCOUNT_EXISTS (email address already exists: alias1@telus-qa-1.eng.zimbra.com)
```
if alias already exists, we cannot make it a primary email.

In succesful case,
`
zimbra@telus-qa-1:~$ zmprov cpe test@telus-qa-1.eng.zimbra.com testnew@telus-qa-1.eng.zimbra.com`

In this case, old primary email will be a new alias that will be added.
<img width="514" alt="image" src="https://github.com/Zimbra/zm-mailbox/assets/104493060/89834c2d-3ec8-4a82-b7b7-68d3368e98a4">

And autcomplete in CLI

```
zimbra@telus-qa-1:~$ zmprov acg telus-qa-1.eng.zimbra.com test
# name uid=testnew,ou=people,dc=telus-qa-1,dc=eng,dc=zimbra,dc=com
createTimeStamp: 20240708054059Z
email: testnew@telus-qa-1.eng.zimbra.com
email2: test@telus-qa-1.eng.zimbra.com
fullName: test
lastName: test
modifyTimeStamp: 20240710124553Z
objectClass: inetOrgPerson
objectClass: zimbraAccount
objectClass: amavisAccount
zimbraId: 35327976-42a8-49e1-949b-c0c5a316bebe

# name uid=testuser,ou=people,dc=telus-qa-1,dc=eng,dc=zimbra,dc=com
createTimeStamp: 20240710045942Z
email: testuser@telus-qa-1.eng.zimbra.com
fullName: testuser
lastName: testuser
modifyTimeStamp: 20240710045942Z
objectClass: inetOrgPerson
objectClass: zimbraAccount
objectClass: amavisAccount
zimbraId: e4bce526-160a-4839-84d1-a649e4b30a3e
```

Essentially, my understanding is primary email would become alias but alias cannot become primary email.

Reply:
Well, if that edge case that I described never actually happens then I guess we don't have to worry about it.

Thank you for having taken a look at it.